tests/test_triton_softmax.py

```diff
@@ -103,3 +103,61 @@ def test_softmax_fp16(dtype):
 
     a = torch.rand(b, s, d, device="cuda", dtype=dtype)
     triton_softmax(a)
+
+
+@pytest.mark.skipif(not _triton_available, reason="Triton is not available")
+@pytest.mark.parametrize("log", [False, True])
+@pytest.mark.parametrize("masking", [True, False])
+@pytest.mark.parametrize("causal", [True, False])
+@pytest.mark.parametrize("contiguous", [True, False])
+def test_softmax_parity_fallback(log, masking, causal, contiguous):
+    """Check that the fallback paths are correct"""
+    torch.random.manual_seed(0)
+
+    shape = (16, 16)
+
+    # Check the result of a FW pass
+    X = torch.normal(0, 1, size=shape, device="cpu", requires_grad=False)
+
+    if not contiguous:
+        # Make sure that the buffer is not contiguous
+        X = X.transpose(-2, -1).contiguous().transpose(-2, -1)
+
+    X_ = X.clone()
+    X.requires_grad = True
+    X_.requires_grad = True
+
+    seq = shape[1]
+    mask = torch.zeros((seq, seq))
+    if masking:
+        mask[torch.rand((seq, seq)) > 0.8] = -float("inf")
+
+    if causal:
+        mask[~torch.tril(torch.ones_like(mask)).bool()] = -float("inf")
+
+    y_torch = (
+        torch.log_softmax(X + mask, dim=-1) if log else torch.softmax(X + mask, dim=-1)
+    )
+    y_triton = (
+        triton_log_softmax(X_, mask, causal)
+        if log
+        else triton_softmax(X_, mask, causal)
+    )
+
+    assert torch.allclose(y_torch, y_triton, equal_nan=True)
+
+    print(y_torch)
```

Comment:
Awesome to have all this testing! Not a bit deal but should we remove some of the print statements here?

Reply:
oh yes sure, sorry about that !